cda-database/src/datatypes/diag_coded_type.rs

```diff
@@ -34,13 +35,822 @@ pub enum DataType {
+
+        if &DataType::Unicode2String == self && !bit_len.is_multiple_of(16) {
+            return Err(DiagServiceError::BadPayload(format!(
+                "Length must be a multiple of 16 bit for {:?}, got {bit_len} bit",
+                &self
+            )));
+        }
+
+        Ok(())
+    }
+}
+
 #[derive(Debug)]
 #[cfg_attr(feature = "deepsize", derive(DeepSizeOf))]
 pub struct DiagCodedType {
-    pub base_datatype: DataType,
-    pub type_: DiagCodedTypeVariant,
+    base_datatype: DataType,
+    type_: DiagCodedTypeVariant,
+    /// Indicates if the byte order is high-low (Big Endian, default) or low-high (Little Endian).
+    is_high_low_byte_order: bool,
 }
-#[derive(Debug)]
+
+enum ByteOrder {
+    /// Byte order is not changed
+    Keep,
+
+    /// Byte order is reversed but acting in pairs of bytes
+    /// Needed for Unicode2String
+    ReorderPairs,
+
+    /// Byte order is reversed
+    Reverse,
+}
+
+impl DiagCodedType {
+    /// Creates a new `DiagCodedType` with high-low byte order.
+    pub fn new_high_low_byte_order(
+        base_datatype: DataType,
+        type_: DiagCodedTypeVariant,
+    ) -> Result<Self, DiagServiceError> {
+        Self::new(base_datatype, type_, true)
+    }
+
+    /// Creates a new `DiagCodedType` with the given byte order.
+    /// Set `is_high_low_byte_order` to `true` for high-low byte order (Big Endian),
+    pub fn new(
+        base_datatype: DataType,
+        type_: DiagCodedTypeVariant,
+        is_high_low_byte_order: bool,
+    ) -> Result<Self, DiagServiceError> {
+        match type_ {
+            DiagCodedTypeVariant::LeadingLengthInfo(_) | DiagCodedTypeVariant::MinMaxLength(_) => {
+                if !matches!(
+                    base_datatype,
+                    DataType::ByteField
+                        | DataType::AsciiString
+                        | DataType::Unicode2String
+                        | DataType::Utf8String
+                ) {
+                    return Err(DiagServiceError::InvalidDatabase(format!(
+                        "LeadingLengthInfo and MinMaxLength are only allowed for ByteField, \
+                         AsciiString, Unicode2String, and Utf8String, got {base_datatype:?}"
+                    )));
+                }
+
+                if let DiagCodedTypeVariant::MinMaxLength(mmlt) = &type_
+                    && base_datatype == DataType::Unicode2String
+                    && (!mmlt.min_length.is_multiple_of(2) || !mmlt.max_length.is_multiple_of(2))
+                {
+                    return Err(DiagServiceError::BadPayload(format!(
+                        "DataType {base_datatype:?} needs an even amount of min/max bytes",
+                    )));
+                }
+            }
+            _ => {}
+        }
+        Ok(Self {
+            base_datatype,
+            type_,
+            is_high_low_byte_order,
+        })
+    }
+
+    pub fn base_datatype(&self) -> &DataType {
+        &self.base_datatype
+    }
```

Comment:
suggestion: as DataType is a small enum that supports `Copy` we can return by value here. This makes the code a bit easier to read in the places this is used because no deref is required then
```suggestion
    pub fn base_datatype(&self) -> DataType {
        self.base_datatype
    }
```